modules/platforms/dotnet/Apache.Ignite.Core/Impl/Unmanaged/NativeLibraryUtils.cs

```diff
@@ -0,0 +1,91 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Apache.Ignite.Core.Impl.Unmanaged
+{
+    using System;
+    using System.Diagnostics;
+    using System.Linq.Expressions;
+    using System.Reflection;
+    using System.Threading;
+
+    /// <summary>
+    /// Native library call utilities.
+    /// </summary>
+    public static class NativeLibraryUtils
+    {
+        /** */
+        private static int _resolversInitialized;
+
+        /// <summary>
+        /// Sets dll import resolvers.
+        /// </summary>
+        public static void SetDllImportResolvers()
+        {
+            if (Interlocked.CompareExchange(ref _resolversInitialized, 1, 0) != 0)
+            {
+                return;
+            }
+
+            // Init custom resolver for .NET 5+ single-file apps.
+            // Do it with Reflection, because SetDllImportResolver is not available on some frameworks,
+            // and multi-targeting is not yet implemented.
+            //
+            // The code below is equivalent to:
+            // NativeLibrary.SetDllImportResolver(typeof(Ignition).Assembly, (libName, _, _) => Resolve(libName));
+            var dllImportResolverType = Type.GetType("System.Runtime.InteropServices.DllImportResolver");
+            var dllImportSearchPathType = Type.GetType("System.Runtime.InteropServices.DllImportSearchPath");
+            var nativeLibraryType = Type.GetType("System.Runtime.InteropServices.NativeLibrary");
+
+            if (dllImportResolverType == null || dllImportSearchPathType == null || nativeLibraryType == null)
+            {
+                return;
+            }
+
+            var setDllImportResolverMethod = nativeLibraryType.GetMethod("SetDllImportResolver");
+
+            if (setDllImportResolverMethod == null)
+            {
+                return;
+            }
+
+            var resolveMethod = typeof(NativeLibraryUtils).GetMethod("Resolve", BindingFlags.Static | BindingFlags.NonPublic);
+            Debug.Assert(resolveMethod != null);
+
+            var libraryName = Expression.Parameter(typeof(string));
+            var assembly = Expression.Parameter(typeof(Assembly));
+            var searchPath = Expression.Parameter(typeof(Nullable<>).MakeGenericType(dllImportSearchPathType));
+
+            var resolve = Expression.Call(resolveMethod, libraryName);
```

Comment:
Maybe change `resolve` to :

```
  Expression<Func<string, IntPtr>> call = s => Resolve(s);
  var resolve = Expression.Invoke(call, libraryName);
```

I think it is a little bit cleaner.

Reply:
Good point, fixed